csi/controller.go

```diff
@@ -980,6 +1008,99 @@ func (s *OsdCsiServer) CreateSnapshot(
 		},
 	}, nil
 }
+func (s *OsdCsiServer) getCloudBackupClient() error {
+	if s.cloudBackupClient == nil {
+		// Get grpc connection
+		conn, err := s.getConn()
+		if err != nil {
+			return status.Errorf(
+				codes.Unavailable,
+				"Unable to connect to SDK server: %v", err)
+		}
+		s.cloudBackupClient = api.NewOpenStorageCloudBackupClient(conn)
+	}
+	return nil
+}
+func (s *OsdCsiServer) createCloudBackup(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
+	err := s.getCloudBackupClient()
+	if nil != err {
+		return nil, err
+	}
+
+	// Get any labels passed in by the CO
+	_, locator, _, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Unable to get parameters: %v", err)
+	}
+
+	backupID := req.GetName()
+	// Create snapshot
+	_, err = s.cloudBackupClient.Create(ctx, &api.SdkCloudBackupCreateRequest{
```

Comment:
If we a return a non-final gRPC error code after this line, will a subsequent s.cloudBackupClient.Create cause issues here? Did you test how that workflow would work?

Reply:
Can you test this implementation in a live PX cluster? I think this will continuously create new snapshots on every retry if it does not exit immediately.